pkg/object/tcpproxy/connection.go

```diff
@@ -0,0 +1,395 @@
+	} else {
+		_ = c.rawConn.(*net.TCPConn).CloseRead()
+	}
+
+	// close rawConn recv, then notify read/write loop to exit
+	close(c.connStopChan)
+	_ = c.rawConn.Close()
+	c.lastBytesSizeRead, c.lastWriteSizeWrite = 0, 0
+
+	if c.onClose != nil {
+		c.onClose(event)
+	}
+	return nil
+}
+
+func (c *Connection) doReadIO() (bufLen int, err error) {
+	if c.readBuffer == nil {
+		c.readBuffer = tcpBufferPool.Get().([]byte)[:iobufferpool.DefaultBufferReadCapacity]
+	}
+
+	// add read deadline setting optimization?
+	// https://github.com/golang/go/issues/15133
+	_ = c.rawConn.SetReadDeadline(time.Now().Add(15 * time.Second))
+	return c.rawConn.(io.Reader).Read(c.readBuffer)
+}
+
+func (c *Connection) doWrite() (int64, error) {
+	_ = c.rawConn.SetWriteDeadline(time.Now().Add(15 * time.Second))
+	bytesSent, err := c.doWriteIO()
+	if err != nil && atomic.LoadUint32(&c.closed) == 1 {
+		return 0, nil
+	}
+
+	if bytesBufSize := c.writeBufLen(); bytesBufSize != c.lastWriteSizeWrite {
+		c.lastWriteSizeWrite = bytesBufSize
+	}
+	return bytesSent, err
+}
+
+func (c *Connection) writeBufLen() (bufLen int) {
+	for _, buf := range c.writeBuffers {
+		bufLen += len(buf)
+	}
+	return
+}
+
+func (c *Connection) doWriteIO() (bytesSent int64, err error) {
+	buffers := c.writeBuffers
+	bytesSent, err = buffers.WriteTo(c.rawConn)
+	if err != nil {
+		return bytesSent, err
+	}
+
+	for i, buf := range c.ioBuffers {
+		c.ioBuffers[i] = nil
+		c.writeBuffers[i] = nil
+		if buf.EOF() {
+			err = iobufferpool.ErrEOF
+		}
+		buf.Release()
+	}
+	c.ioBuffers = c.ioBuffers[:0]
+	c.writeBuffers = c.writeBuffers[:0]
+	return
+}
+
+// ServerConnection wrap tcp connection to backend server
+type ServerConnection struct {
+	Connection
+	connectTimeout time.Duration
+	connectOnce    sync.Once
+}
+
+// NewServerConn construct tcp server connection
+func NewServerConn(connectTimeout uint32, serverAddr net.Addr, listenerStopChan chan struct{}) *ServerConnection {
+	conn := &ServerConnection{
+		Connection: Connection{
+			remoteAddr: serverAddr,
+
+			writeBufferChan: make(chan *iobufferpool.StreamBuffer, writeBufSize),
+
+			mu:               sync.Mutex{},
+			connStopChan:     make(chan struct{}),
+			listenerStopChan: listenerStopChan,
+		},
+		connectTimeout: time.Duration(connectTimeout) * time.Millisecond,
+	}
+	return conn
+}
+
+// Connect create backend server tcp connection
+func (u *ServerConnection) Connect() (err error) {
+	u.connectOnce.Do(func() {
+		addr := u.remoteAddr
+		if addr == nil {
+			err = errors.New("server addr is nil")
+			return
+		}
+
+		timeout := u.connectTimeout
+		if timeout == 0 {
+			timeout = 10 * time.Second
+		}
+
+		u.rawConn, err = net.DialTimeout("tcp", addr.String(), timeout)
+		if err != nil {
+			if err == io.EOF {
+				err = errors.New("server has been closed")
+			} else if te, ok := err.(net.Error); ok && te.Timeout() {
+				err = errors.New("connect to server timeout")
+			} else {
+				err = errors.New("connect to server failed")
+			}
```

Comment:
do we need to distinguish these errors and overwrite the low-level errors?

Reply:
No need, I have corrected it in the new commit.